### PR TITLE
Update design and add EventsQueue's types info

### PR DESCRIPTION
**Pull request by mauropasse**
Signed-off-by: Mauro Passerino <mpasserino@irobot.com>

### CODE REVIEW

File: articles/events_executor.md [hunk 1]
```diff
@@ -53,33 +53,36 @@ The `EventsExecutor` is characterized by two main components, i.e. the `EventsQu
...
 ### Brief Overview
 
-In order to execute generic ROS 2 entities (i.e. everything except timers), the `EventsExecutor` relies on an event queue.
+In order to execute generic ROS 2 entities, the `EventsExecutor` relies on an event queue.
 Whenever a ROS 2 entity that is associated to the executor has work to do, it will push an event into the executor's queue.
 Then the executor can process these events in a FIFO manner, without need for expensive entities look-ups.
 Processing an event results in different operations depending on the entity that generated it.
 
-Timers are executed in a separate task by a timers manager, where they are kept in a priority queue sorted according to their expiration time. The task then has to monitor only the first element of the queue and can execute its callback as soon as it expires.
+Timers are monitored in a separate task by a timers manager, where they are kept in a priority queue sorted according to their expiration time. The task then has to monitor only the first element of the queue and can execute its callback as soon as it expires. The timers manager can also push timer events into the queue, instead of executing the timer.
```
Comment (alsora) [inline articles/events_executor.md]: General comment: go to a new line after a `.`

File: articles/events_executor.md [hunk 2]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
 Whenever an entity is associated to an executor it may already have multiple items of work ready to be processed.
 The `EventsExecutor` should tell the entity whether it has to immediately push or to discard those pending events.
 
+### Types of Events queue
+
+There are multiple uses cases with different requirements (performance, determinism in events ordering, bounded memory). A single `EventsQueue` can't comply with all requirements simultaneously, so there's need for different types of queue.
```
Comment (alsora) [inline articles/events_executor.md]: We should focus on describing the design of the base class.

File: articles/events_executor.md [hunk 3]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
...
+
+##### SimpleEventsQueue
+
+This is the simplest queue, as it does not peform any checks when events are pushed or extracted, so it's more suitable for situacions where CPU performance is needed.
```
Comment (alsora) [inline articles/events_executor.md]: I would just say that this is the simplest queue, just showing how to wrap std::queue into the api.

it should never be used as it can grow unbounded

File: articles/events_executor.md (hunk 2 above)
Comment (alsora) [reply articles/events_executor.md]: let's have a small paragraph here listing the possible type of queues that could be implemented

File: articles/events_executor.md [hunk 4]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
...
+ - Event 1 from Subscription A
+ - Event 2 from Subscription B
+ - Event 3 from Subscription A
+
+When the `EventsExecutor` is ready to process these new events, it will start from the first pushed, i.e. the event 1 from Subscription A.
+However, when calling the `execute_subscription()` API, the message that generated event 1 will not be available anymore as it has been overwritten from the message that generated event 3.
+The `execute_subscription()` API will not fail, but rather it will process the message that generated event 3.
+This violates the ordering as event 2 should have been processed before that.
+
+##### LockFreeEventsQueue
```
Comment (alsora) [inline articles/events_executor.md]: let's remove this section and we can just mention above that if needed you can implement a lock free queue

File: articles/events_executor.md [hunk 5]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
...
+
+##### BoundedEventsQueue
+
+This queue doesn't allow more events from an entity than its history size. For example a subscription with a history size of 5, can't have more than 5 events from it in the queue.
+
+This queue has policies to decide what to do when a new event arrives from an entity which will exceed the amount of events allowed. It can remove the oldest event and push a new one, which keeps the relative time ordering of the event with respect to other events (at the cost of some CPU time), or it can directly avoid pushing the new event into the queue, saving CPU but subverting the time ordering of events.
+
+##### WaitSetEventsQueue
+
+This queue has a waitset-like behaviour: There's a single entry for each entity, and the order of events execution is the same as the `SingleThreadedExecutor` executor.
```
Comment (alsora) [inline articles/events_executor.md]: typo: `There` -> `there`.
it's unclear what `single entry for each entity` means.
this is a design document so we should describe the intended behavior, not referrign to a different executor.

File: articles/events_executor.md [hunk 6]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
...
+If the `TimersManager` is configured to push timer events, timers will be executed in the same thread as other entities, as it happens also with the current default executor.
```
Comment (alsora) [inline articles/events_executor.md]: This is unrelated to the queue

File: articles/events_executor.md [hunk 7]
```diff
@@ -129,11 +132,63 @@ This class will forward the items needed for pushing events to the notify guard
...
+
+The difference with respect to the waitset used in the `SingleThreadedExecutor` is that only entities which have work to do are present in the waitset. This way we avoid polling entities in search of new data to process.
+Also this reduces the time to look for existing elements in the waitset (to update their events counter) as there are less elements to iterate.
+
+The events from a single entity are represented by a single event and counter, specifying the number of events left to execute from this entity (as oposed to other EventsQueue which can have multiple events).
+
+The waitset has elements ordered by entity tipe: 1.Timers / 2.Subs / 3.Services / 4.Clients / 5.Waitables. An iterator points to the next event to dequeue, so the order of execution of entities is the same as of the `SingleThreadedExecutor`.
```
Comment (alsora) [inline articles/events_executor.md]: all these are implementation details, they are not relevant

File: articles/events_executor.md (hunk 1 above)
Comment (mauropasse) [reply articles/events_executor.md]: Done in b73651f

File: articles/events_executor.md (hunk 7 above)
Comment (mauropasse) [reply articles/events_executor.md]: Removed in 0d70bf2

File: articles/events_executor.md (hunk 6 above)
Comment (mauropasse) [reply articles/events_executor.md]: Removed in 0d70bf2

File: articles/events_executor.md (hunk 5 above)
Comment (mauropasse) [reply articles/events_executor.md]: Updated in 0d70bf2

File: articles/events_executor.md (hunk 4 above)
Comment (mauropasse) [reply articles/events_executor.md]: Removed and menitioned in 0d70bf2

File: articles/events_executor.md (hunk 3 above)
Comment (mauropasse) [reply articles/events_executor.md]: Done in 0d70bf2

File: articles/events_executor.md (hunk 2 above)
Comment (mauropasse) [reply articles/events_executor.md]: Done in 0d70bf2
